Added analytics - telemetry example

papafe opened the pull request:
Not sure if we want to add the example app in the repo here, but I thought this would make it easier to review. 

Fixes #3324 

##  TODO

* [x] List of examples in the main README (?)
* [x] Export charts file from dashboard and add link to instructions on how to import it (https://www.mongodb.com/docs/charts/dashboards/#import-a-dashboard-from-a-file)

nirinchev commented on examples/AnalyticsTelemetrics/README.md:
```diff
@@ -0,0 +1,101 @@
+# AnalyticsTelemetry #
+
+This example MAUI application shows how to use Realm for Analytics or Telemetry application, by using [Data Ingest](https://www.mongodb.com/docs/realm/sdk/dotnet/sync/asymmetric-sync/), and how to visualize the captured data using [Charts](https://www.mongodb.com/docs/charts/). Data Ingest can be used when you need to sync data unidirectionally to Atlas, for example with telemetry in IoT applicat ...
```
```suggestion
This example MAUI application shows how to use Realm for Analytics or Telemetry application, by using [Data Ingest](https://www.mongodb.com/docs/realm/sdk/dotnet/sync/asymmetric-sync/), and how to visualize the captured data using [MongoDB Charts](https://www.mongodb.com/docs/charts/). Data Ingest can be used when you need to sync data unidirectionally to Atlas, for example with telemetry in IoT applications, or you need to write immutable data and do not need conflict resolution, for example when logging analytics events. 
```

nirinchev commented on examples/AnalyticsTelemetrics/README.md:
```diff
@@ -0,0 +1,101 @@
+{
+    [MapTo("_id")]
+    [PrimaryKey]
+    public ObjectId Id { get; set; } = ObjectId.GenerateNewId();
```
```suggestion
    public ObjectId Id { get; private set; } = ObjectId.GenerateNewId();
```

papafe replied:
Done

nirinchev commented on examples/AnalyticsTelemetrics/Resources/Fonts/OpenSans-Regular.ttf:
Do we need these fonts?

nirinchev commented on examples/AnalyticsTelemetrics/Resources/Images/dotnet_bot.svg:
Do we need this image?

nirinchev commented on examples/AnalyticsTelemetrics/Resources/Raw/AboutAssets.txt:
Can probably be deleted

nirinchev commented on examples/AnalyticsTelemetrics/Resources/Splash/splash.svg:
Should we create branded icon/splash screen?

nirinchev commented on examples/AnalyticsTelemetrics/ViewModels/AnalyticsPageViewModel.cs:
```diff
@@ -0,0 +1,174 @@
+
+            public AnalyticsData GetRandomAnalyticsEvent()
+            {
+                return new AnalyticsData()
```
```suggestion
                return new()
```

nirinchev commented on examples/AnalyticsTelemetrics/ViewModels/TelemetryPageViewModel.cs:
```diff
@@ -0,0 +1,150 @@
+
+            public TemperatureReading GetTemperatureReading()
+            {
+                return new TemperatureReading()
```
```suggestion
                return new()
```

nirinchev reviewed (commented):
Maybe rename the example to AnalyticsTelemetry? I guess the benefit of telemetrics is that it kinda rhymes with analytics, but I don't think it's widely used terminology for what we're doing here.

papafe commented on examples/AnalyticsTelemetrics/ViewModels/TelemetryPageViewModel.cs:
```diff
@@ -0,0 +1,150 @@
+
+                    AddToLog("- - - - - -");
+                    Console.WriteLine(Environment.CurrentManagedThreadId);
+                    await Task.Delay(_collectionDelayMilliseconds, cancellationToken);
```
Probably I would go with number 1, it seems to be the "easiest" solution.

papafe replied:
I'm actually starting to think that number 3 could be the best option. `Task.Delay().Wait()` looks a little strange, and then we don't even get the "clean" `TaskCanceledException` but an `Aggregate` one. 
I will add comments on why we do that, and that if this is run always on the same thread we can move it out of the loop.

nirinchev replied:
I'd probably go for async context solution - the main reason is that if people copy-paste it, they'll be able to add more async stuff in that loop - e.g. `await sensor.GetInfo()` and things like that.

papafe replied:
Makes sense.

papafe commented on examples/AnalyticsTelemetrics/ViewModels/TelemetryPageViewModel.cs:
```diff
@@ -0,0 +1,150 @@
+                {
+                    foreach (var sensor in _sensors)
+                    {
+                        var realm = RealmService.GetRealm();
```
I think you saw it down here, but it's because I'm awaiting asynchronously on Task.Delay, so we could switch thread after that. But I'll change that part so we don't need this. And yes, I'll add the using

papafe commented on examples/AnalyticsTelemetrics/README.md:
```diff
@@ -0,0 +1,101 @@
+
+Aside from the aforementioned rules, asymmetric objects can be created and added to a realm in the same way as any other realm object. 
+
+## Configuration ##
```
I think it would make more sense to link to the Microsoft documentation for the installation, so we're sure that they're always updated.

nirinchev replied:
I think we can still provide the minimum setup commands to avoid having to read an entire page of documentation to do something simple. We can still link to the MS docs page for users that need more details, but - for example - if I'm a WPF developer and don't know too much about MAUI, it should still be easy for me to get the app up and running and see something happening.

papafe replied:
Done

papafe commented on examples/AnalyticsTelemetrics/README.md:
```diff
@@ -0,0 +1,101 @@
+
+    public DateTimeOffset Timestamp { get; set; }
+
+    public string EventType { get; set; }
```
The default maui project doesn't have nullability enabled, so I forgot to enable it :)

nirinchev replied:
I think we should enable it - not sure why it's not on by default for MAUI, but we did spend quite a bit of time to support nullability, it'd be a shame if we don't showcase it at least in the sample apps 😄

papafe replied:
Definitely! It's coming 😁

papafe replied:
Done

papafe commented on examples/AnalyticsTelemetrics/Services/RealmService.cs:
```diff
@@ -0,0 +1,37 @@
+
+        public static async Task Init()
+        {
+            if (_serviceInitialised)
+            {
+                return;
+            }
```
This is called only from `OnAppearing` in the MainPage from an async command, so it should not execute if the method has not completed, even if the user is going back and forth between pages. 
Yes, this could cause a race condition if not used correctly, but I thought that it would make it more complicated to read.

nirinchev replied:
Gotcha, I agree it's probably not worth complicating it.

papafe commented on examples/AnalyticsTelemetrics/README.md:
```diff
@@ -0,0 +1,101 @@
+
+## Analytics ##
+
+The analytics part of the application simulates the generation of analytics events, using the `AnalyticsData` asymmetric object model:
```
Done

papafe commented:
@nirinchev I think I more or less incorporated all your comments:
- Running the generation tasks inside `AsyncContext.Run` and added comments on why we do it
- Added more info in the README files about what we are trying to show
- Added also the `.charts` files from the dashboards, so that users can just import them
- Removed the default fonts/images and copied the theme from our template app

There are only a couple of things left:
- For some reason (probably because of https://github.com/dotnet/maui/issues/14421), on iOS (only) the splash screen doesn't appear and there is a black screen. Hopefully I can find a workaround (??)
- Modify the readme of the repo to add a line about this example
- Change the name of the solution. I want to do this at the end, so it's easier to read for now.

papafe commented:
@nirinchev The only thing left for me is to try to fix the issue with the splash screen on iOS, but I've finished all the things I've wrote before, so it's very close to being mergeable for me

papafe commented:
I've just spent 30 min trying to fix the black splash screen to no avail. I copied everything from the template app, where the splash screen was working, but the only thing I've managed to achieve, is that now it doesn't work in the template app either 🙃 
Given that there is already a ticket opened for this, I would avoid spending more time on this, so for me this can be merged

nirinchev reviewed (approved):
Looks good - I think we should probably move it to https://github.com/realm/realm-dotnet-samples similarly to the swift example.

papafe commented:
Closing this PR has this has been added to our example repo
